scipy/spatial/transform/_rigid_transformation.pyx

```
@@ -0,0 +1,1978 @@
+        # where R is a 3x3 orthonormal rotation matrix and t is a 3x1 translation
+        # vector. The last row is always [0, 0, 0, 1] exactly
+
+        # Check the last row. Exact match is required, as this last row should
+        # not accumulate floating point error during composition.
+        last_row_ok = np.all(matrix[:, 3, :] == np.array([0, 0, 0, 1]), axis=1)
+
+        # Check the determinant of the rotation matrix
+        # (should be positive for right-handed rotations)
+        dets = np.linalg.det(matrix[:, :3, :3])
+
+        # Gramian orthonormality check
+        # (should be the identity matrix for orthonormal matrices)
+        grams = matrix[:, :3, :3] @ np.swapaxes(matrix[:, :3, :3], 1, 2)
+        orthonormal_ok = np.all(np.isclose(grams, np.eye(3)), axis=(1, 2))
+
+        for ind in range(num_transformations):
+            if not last_row_ok[ind]:
+                raise ValueError(f"Expected last row of transformation matrix {ind} to be "
+                                 f"exactly [0, 0, 0, 1], got {matrix[ind, 3, :]}.")
+
+            # Check that the rotation matrix is orthonormal
+            if not np.isclose(dets[ind], 1) or not orthonormal_ok[ind]:
+                if normalize:
+                    if dets[ind] <= 0:
+                        raise ValueError("Non-positive determinant (left-handed or null "
+                                         f"coordinate frame) in rotation component of "
+                                         f"transformation matrix {ind}: {matrix[ind]}.")
+                    else:
+                        # Orthonormalize the rotation matrix
+                        U, _, Vt = np.linalg.svd(matrix[ind, :3, :3])
+                        matrix[ind, :3, :3] = U @ Vt
+                else:
+                    raise ValueError("Expected rotation component of transformation "
+                                     f"matrix {ind} be orthonormal: {matrix[ind]}.")
+
+        self._matrix = matrix
```

I still would consider the approach of `Rotation.from_matrix`, there would be consistency between them.

Nothing is wrong with that code though.

---

It seems somewhat messy to me to cast from a matrix to a quaternion and then back. And personally I'm not familiar enough with the algorithms in Rotation.from_matrix and to_matrix to  know how to handle the normalization check.

---

The reference to the paper is in `Rotation.from_matrix` documentation. The emphasis of the author is that the method is very efficient and close to the Procrustes method you've implemented in terms of accuracy (the details are somewhat complicated).

From programming perspective reusing robust `Rotation` doesn't look like a bad idea too.

---

Hmmm, it looks like switching from the svd method to `Rotation.from_matrix(m).as_matrix()` is producing significantly different results. That's a little worrisome, I'll have to dig into the paper.

```
        # Test non-orthogonal rotation matrix
        matrix = np.array([[1, 1, 0, 0],
                           [0, 1, 0, 0],
                           [0, 0, 1, 0],
                           [0, 0, 0, 1]])
        tf = RigidTransformation.from_matrix(matrix)
        expected = np.array([[0.894427,  0.447214, 0, 0],
                             [-0.447214,  0.894427, 0, 0],
                             [0, 0, 1, 0],
                             [0, 0, 0, 1]])
>       assert_allclose(tf.as_matrix(), expected, atol=1e-6)
E       AssertionError: 
E       Not equal to tolerance rtol=1e-07, atol=1e-06
E       
E       Mismatched elements: 4 / 16 (25%)
E       Max absolute difference among violations: 0.02337424
E       Max relative difference among violations: 0.05226633
E        ACTUAL: array([[ 0.882353,  0.470588,  0.      ,  0.      ],
E              [-0.470588,  0.882353, -0.      ,  0.      ],
E              [-0.      ,  0.      ,  1.      ,  0.      ],
E              [ 0.      ,  0.      ,  0.      ,  1.      ]])
E        DESIRED: array([[ 0.894427,  0.447214,  0.      ,  0.      ],
E              [-0.447214,  0.894427,  0.      ,  0.      ],
E              [ 0.      ,  0.      ,  1.      ,  0.      ],
E              [ 0.      ,  0.      ,  0.      ,  1.      ]])
```

---

Reading the paper, it seems like the algorithm is only valid for near-orthogonal matrices. That's fine for normalizing round-off error but I think is probably not a good algorithm for initializing transformations.

I don't think this should be the default behavior for Rotations either. In the `Rotation.from_matrix()` docs, we have this example:
```
        If input matrices are not special orthogonal (orthogonal with
        determinant equal to +1), then a special orthogonal estimate is stored:

        >>> a = np.array([
        ... [0, -0.5, 0],
        ... [0.5, 0, 0],
        ... [0, 0, 0.5]])
        >>> np.linalg.det(a)
        0.12500000000000003
        >>> r = R.from_matrix(a)
        >>> matrix = r.as_matrix()
        >>> matrix
        array([[-0.38461538, -0.92307692,  0.        ],
               [ 0.92307692, -0.38461538,  0.        ],
               [ 0.        ,  0.        ,  1.        ]])
        >>> np.linalg.det(matrix)
        1.0000000000000002
```

The matrix we get there is far off from the `[[0, -1, 0], [1, 0, 0], [0, 0, 1]]` matrix that should return from an orthonormalization process of that input matrix, which is what the svd method returns.

```python
import numpy as np
a = np.array([
    [0, -0.5, 0],
    [0.5, 0, 0],
    [0, 0, 0.5]])
U, _, Vt = np.linalg.svd(a)
print(U @ Vt)
```
```python
[[ 0. -1.  0.]
 [ 1.  0.  0.]
 [ 0.  0.  1.]]
```

---

Created an issue here: https://github.com/scipy/scipy/issues/22417

---

I've vectorized this code block a bit for speed, but kept the svd method.

---

PR for fixing rotations: https://github.com/scipy/scipy/pull/22418

---

As a side note, I recently found out that there is yet another orthonormalization method: https://zalo.github.io/blog/polar-decomposition/

The article also presents a simpler Gram-Schmidt orthonormalization.

---

Agree that we should switch to proper orthgonalization then. As for the method - anything which can be easily implement in scipy stack (SVD looks more than fine in this regard).

---

The polar decomposition looks interesting, here's the paper that page cites: https://animation.rwth-aachen.de/media/papers/2016-MIG-StableRotation.pdf

I don't love that it's an iterative formula with a convergence tolerance, which makes sense for computer graphics but not as much for scientific programming.

---

I do like that it is able to robustly handle rank deficient and left handed matrices. IMO I would rather raise an error for those inputs (as introduced in the above PR), but there's an argument to be made to always try and best-fit an arbitrary matrix to a right-handed orthogonal one. Thoughts?

This is in contrast to SVD which might otherwise return a left-handed rotation matrix that needs to be checked. See for example: https://github.com/pyvista/pyvista/blob/main/pyvista/core/utilities/transformations.py#L340-L519

---

Updated to use `Rotation.from_matrix`, which now correctly normalizes inputs.